compiler/src/iree/compiler/Dialect/HAL/Transforms/InitializeDevices.cpp

```diff
@@ -0,0 +1,111 @@
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir::iree_compiler::IREE::HAL {
+
+#define GEN_PASS_DEF_INITIALIZEDEVICESPASS
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h.inc"
+
+namespace {
+
+// Converts an initialized device global to one with a util.initializer that
+// performs the device initialization. The initializer is added immediately
+// following the global in its parent op.
+static void initializeDeviceGlobal(
+    IREE::Util::GlobalOpInterface globalOp,
+    IREE::HAL::DeviceInitializationAttrInterface initialValue,
+    const IREE::HAL::TargetRegistry &targetRegistry) {
+  auto loc = globalOp.getLoc();
+
+  // Clear the initial value as we'll be initializing from the initializer.
+  globalOp.setGlobalInitialValue({});
+
+  // Build a new util.initializer.
+  OpBuilder moduleBuilder(globalOp);
+  moduleBuilder.setInsertionPointAfter(globalOp);
+  auto initializerOp = moduleBuilder.create<IREE::Util::InitializerOp>(loc);
+  auto *block = moduleBuilder.createBlock(&initializerOp.getBody());
+  auto initializerBuilder = OpBuilder::atBlockBegin(block);
+
+  // Get the device from the attribute builder; note that it may be null.
+  Value enumeratedDevice = initialValue.buildDeviceEnumeration(
+      loc,
+      [&](Location loc, Value device, IREE::HAL::DeviceTargetAttr targetAttr,
+          OpBuilder &builder) {
+        auto targetDevice =
+            targetRegistry.getTargetDevice(targetAttr.getDeviceID());
+        return targetDevice ? targetDevice->buildDeviceTargetMatch(
+                                  loc, device, targetAttr, builder)
+                            : Value{};
+      },
+      initializerBuilder);
+
+  // Check if the device is null and error out. We could support optional
+  // devices that are allowed to be null but don't support that anywhere else in
+  // the compiler today and may never want to. If selecting from multiple
+  // devices queries can be used to detect what the selected device was and
+  // those will be memoized.
+  Value nullDevice = initializerBuilder.create<IREE::Util::NullOp>(
+      loc, enumeratedDevice.getType());
+  Value isNull = initializerBuilder.create<IREE::Util::CmpEQOp>(
+      loc, enumeratedDevice, nullDevice);
+  initializerBuilder.create<scf::IfOp>(
+      loc, isNull, [&](OpBuilder &thenBuilder, Location thenLoc) {
+        Value status = thenBuilder.create<arith::ConstantIntOp>(
+            thenLoc, static_cast<int64_t>(IREE::Util::StatusCode::NotFound),
+            32);
+        std::string str;
+        {
+          llvm::raw_string_ostream os(str);
+          os << "HAL device `" << globalOp.getGlobalName().getValue()
+             << "` not found or unavailable: ";
+          initialValue.printStatusDescription(os);
+        }
+        thenBuilder.create<IREE::Util::StatusCheckOkOp>(thenLoc, status, str);
+        thenBuilder.create<scf::YieldOp>(thenLoc);
+      });
```

Comment:
Do we have a helper function for creating a status check like this? Want one?
```
// CHECK-NEXT: scf.if %[[IS_NULL]] {
//      CHECK:   util.status.check_ok %c5_i32, "HAL device `device_fallback` not found or unavailable: #hal.device.fallback<@device_base>"
```

something like
`createCheckOkIfCondition(loc, isNull, str);`

Reply:
I've been meaning to clean up that op for awhile - I should do that - also feeling the same with queue affinities. I really like the enum ops I added in the HAL and could do something similar:
```mlir
%not_found = util.status.code<NotFound> : i32
util.status.check %not_found, "message"
```
I've also wanted to do this with queue affinities - instead of `= arith.constant 3 : i64` being "queues 0 and 1" it'd be nice to have `= hal.affinity<[0, 1]>`.

Filed an issue to track cleaning these up (I'll do it after all this lands): #17941